android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/RoutingSource.kt

```diff
@@ -0,0 +1,190 @@
+import com.badoo.ribs.core.routing.history.RoutingHistoryElement.Activation.INACTIVE
+import com.jakewharton.rxrelay2.BehaviorRelay
+import com.jakewharton.rxrelay2.Relay
+import io.reactivex.Observable
+import io.reactivex.ObservableSource
+import io.reactivex.Observer
+import io.reactivex.functions.BiFunction
+import java.security.PrivateKey
+
+interface RoutingSource<C : Parcelable> :
+    ObservableSource<RoutingHistory<C>>,
+    SubtreeBackPressHandler {
+
+    operator fun plus(other: RoutingSource<C>): RoutingSource<C> = Combined(this, other)
+
+    fun remove(identifier: Identifier)
+
+    data class Combined<C : Parcelable>(
+        val first: RoutingSource<C>,
+        val second: RoutingSource<C>
+    ) : RoutingSource<C> {
+
+        data class CombinedHistory<C : Parcelable>(
+            val first: RoutingHistory<C>,
+            val second: RoutingHistory<C>
+        ):  RoutingHistory<C> {
+
+            override fun iterator(): Iterator<RoutingHistoryElement<C>> =
+                ConcatIterator(first.iterator()) + second.iterator()
+        }
+
+        private val combined = Observable.combineLatest(
+            first,
+            second,
+            BiFunction<RoutingHistory<C>, RoutingHistory<C>, RoutingHistory<C>> { source1, source2 ->
+                CombinedHistory(source1, source2)
+            }
+        )
+
+        override fun subscribe(observer: Observer<in RoutingHistory<C>>) {
+            combined.subscribe(observer)
+        }
+
+        override fun remove(identifier: Identifier) {
+            first.remove(identifier)
+            second.remove(identifier)
+        }
+
+        override fun handleBackPressFirst(): Boolean =
+            first.handleBackPressFirst() || second.handleBackPressFirst()
+
+        override fun handleBackPressFallback(): Boolean =
+            first.handleBackPressFallback() || second.handleBackPressFallback()
+    }
+
+    // TODO extract
+    class Permanent<C : Parcelable>(permanents: Iterable<C>) : RoutingSource<C> {
+
+        companion object {
+            fun <C : Parcelable> permanent(permanents: Iterable<C>) = Permanent(permanents)
+
+            fun <C : Parcelable> permanent(vararg permanents: C) = Permanent(permanents.toSet())
+        }
+
+        private val routingElements =
+            permanents.mapIndexed { idx, configuration ->
+                RoutingHistoryElement(
+                    activation = ACTIVE,
+                    routing = Routing(
+                        configuration = configuration,
+                        identifier = Identifier("Permanent $idx")
+                    )
+            ) }
+
+        private val permanentHistory =
+            RoutingHistory.from(routingElements)
+
+        override fun remove(identifier: Identifier) {
+            // no-op -- it's permanent!
+        }
+
+        override fun subscribe(observer: Observer<in RoutingHistory<C>>) {
+            Observable
+                .just(permanentHistory)
```

Comment:
Is it really needed to create Observable each time?

Reply:
good catch, will fix